test/e2e/background-socket/window-handles.ts

```diff
@@ -0,0 +1,208 @@
+  ) {
+    const currentHandle =
+      optionalCurrentHandle || (await this.driver.getWindowHandle());
+
+    let currentTitle, currentUrl;
+
+    // Wait 25 x 200ms = 5 seconds for the title and url to be set
+    for (let i = 0; i < 25 && !currentTitle && !currentUrl; i++) {
+      currentTitle = await this.driver.getTitle();
+      currentUrl = await this.driver.getCurrentUrl();
+
+      if (!currentTitle || !currentUrl) {
+        await this.driver.sleep(200);
+      }
+    }
+
+    if (!currentTitle || !currentUrl) {
+      log.debug(
+        'Cannot get properties of current window',
+        currentHandle,
+        currentTitle,
+        currentUrl,
+      );
+      return null;
+    }
+
+    let annotatedHandle = this.annotatedHandles.find(
+      (a) => a.id === currentHandle,
+    );
+
+    if (annotatedHandle) {
+      // Handle already there, update it
+      annotatedHandle.title = currentTitle;
+      annotatedHandle.url = currentUrl;
+    } else {
+      // Handle not there, add it
+      annotatedHandle = {
+        id: currentHandle,
+        title: currentTitle,
+        url: currentUrl,
+      };
+      this.annotatedHandles.push(annotatedHandle);
+    }
+
+    if (property) {
+      // Return the current title or url
+      return annotatedHandle[property];
+    }
+
+    // Didn't ask for any return value
+    return null;
+  }
+
+  // Count the number of handles with an unknown title
+  countUntitledHandles() {
+    return this.annotatedHandles.filter((a) => !a.title).length;
+  }
+
+  // Count the number of handles with an unknown template property
+  countHandlesWithoutProperty(property: WindowProperties) {
+    return this.annotatedHandles.filter((a) => !a[property]).length;
+  }
+
+  /**
+   * Switches the context of the browser session to the window/tab with the given property.
+   *
+   * You can think of this kind of like a template function:
+   * If `property` is `title`, then this becomes `switchToWindowWithTitle`
+   * If `property` is `url`, then this becomes `switchToWindowWithUrl`
+   * Remember that `a[property]` becomes `a.title` or `a.url`
+   *
+   * @param property - 'title' or 'url'
+   * @param value - The value we're searching for and want to switch to
+   * @returns The handle of the window tab with the given property value
+   */
+  async switchToWindowWithProperty(property: WindowProperties, value: string) {
+    // Ask the extension to wait until the window with the given property is open
+    // (just ignore the return value here, because the tabs are not in the same order as driver.getAllWindowHandles())
+    await getServerMochaToBackground().waitUntilWindowWithProperty(
+      property,
+      value,
+    );
+
+    await this.getAllWindowHandles();
+
+    // See if we already know the handle by annotation
+    const handle = this.annotatedHandles.find((a) => a[property] === value);
+    let handleId = handle?.id;
+
+    // If there's exactly one un-annotated handle, we can use it
```

Comment:
Why is it safe to assume this is our window?

Reply:
The way we always call it, it would be, but in the larger sense, I guess it's not.  We should explicitly get the title/url.

Reply:
We now assume it's our window, but then verify